middleware/http_cluster.go

```diff
@@ -0,0 +1,148 @@
+package middleware
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/dskit/clusterutil"
+	"github.com/grafana/dskit/httpgrpc"
+)
+
+type clusterValidationError struct {
+	ClusterValidationErrorMessage string `json:"cluster_validation_error_message"`
+}
+
+// writeAsJSON writes this error as JSON to the HTTP response.
+func (e *clusterValidationError) writeAsJSON(w http.ResponseWriter) {
+	data, err := json.Marshal(e)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusNetworkAuthenticationRequired)
+	// We ignore errors here, because we cannot do anything about them.
+	// Write will trigger sending Status code, so we cannot send a different status code afterwards.
+	// Also, this isn't internal error, but error communicating with client.
+	_, _ = w.Write(data)
+}
+
+func ClusterValidationRoundTripper(cluster string, invalidClusterValidationReporter InvalidClusterValidationReporter, next http.RoundTripper) RoundTripperFunc {
+	validateClusterValidationRoundTripperInputParameters(cluster, invalidClusterValidationReporter)
+	return func(req *http.Request) (*http.Response, error) {
+		clusterutil.PutClusterIntoHeader(req, cluster)
+		resp, err := next.RoundTrip(req)
+		if err != nil {
+			return nil, err
+		}
+		if resp.StatusCode != http.StatusNetworkAuthenticationRequired {
+			return resp, nil
+		}
+		if resp.Header.Get("Content-Type") != "application/json" {
+			return resp, nil
+		}
+		defer resp.Body.Close()
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		var clusterValidationErr clusterValidationError
+		err = json.Unmarshal(body, &clusterValidationErr)
+		if err != nil {
+			resp.Body = io.NopCloser(bytes.NewReader(body))
+			return resp, nil
```

Comment:
Since the server returned an unexpected format (and we don't necessarily read the whole body), I think we should return internal server error (it's a scenario that should not happen):
```suggestion
			resp.Body = nil
			resp.StatusCode = http.InternalServerError
			resp.Header.Set("Content-Type", "text/plain")
			return resp, nil
```

Reply:
I think this would be unexpected and hard to debug if happens. 

Imagine that you're seeing lots of internal errors on the client but server doesn't send any?

Let's just return something like `fmt.Errorf("server returned 511 but the result wasn't a valid json: %w", err)`,